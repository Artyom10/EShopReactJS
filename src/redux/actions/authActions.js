export const signIn = (credentials) => {
    debugger;
    return (dispatch, getState, {getFirebase}) => {
        const firebase = getFirebase();

        firebase.auth().signInWithEmailAndPassword(
            credentials.email,
            credentials.password
        ).then(() => {
            dispatch({type: 'LOGIN_SUCCESS'});
        }).catch((err) => {
            dispatch({type: 'LOGIN_ERROR', err})
        });
    }
};

export const signOut = () => {
    return(dispatch, getState, {getFirebase}) => {
        const firebase = getFirebase();

        firebase.auth().signOut().then(() => {
            dispatch({type: 'SIGNOUT_SUCCESS'})
        });
    } 
}

export const signUp = (newUser) => {
    return(dispatch, getState, {getFirebase, getFirestore}) => {
        debugger;
        const firebase = getFirebase();
        const firestore = getFirestore();

        firebase.auth().createUserWithEmailAndPassword(
            newUser.email,
            newUser.password,
        ).then((response) => {
            return firestore.collection('users').doc(response.user.uid).set({
                firstName: newUser.firstName,
                secondName: newUser.secondName,
                request: false, //newUser.request,
                bags: [], //newUser.bags,
                userPhoto: 'https://icons-for-free.com/iconfiles/png/512/business+costume+male+man+office+user+icon-1320196264882354682.png',
                valuedProducts: [],//newUser.valuedProducts,
                isAdmin: false,
                isWho: 'user',
                email: newUser.email,
            })
        }).then(() => {
            dispatch({type: 'SIGNUP_SUCCESS'})
        }).catch(err => {
            dispatch({type: 'SIGNUP_ERROR', err})
        })
    }
}