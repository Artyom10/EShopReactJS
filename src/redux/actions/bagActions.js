/*export const removeFromBag = (removeTargetBag) => {
    return (dispatch, getState) => {
 dispatch({ type: 'REMOVE_FROM_BAG', removeTargetBag,})
    }
}*/

export const addToBag = (targetAddToBag) => {
    return (dispatch, getState,{getFirebase, getFirestore}) => {
     const firestore = getFirestore();
     //const profile = getState()
     const buyerId = getState().firebase.auth.uid
     firestore.collection('users').doc(buyerId).update({
            bags: [targetAddToBag]
     })
     .then(() => {
         dispatch({type: 'ADD_TO_BAG', targetAddToBag})
     }).catch((err) => {
         dispatch({type: 'ADD_TO_BAG_ERROR', err})
     })
    }
}


 //dispatch({type: 'ADD_TO_BAG', targetAddToBag})