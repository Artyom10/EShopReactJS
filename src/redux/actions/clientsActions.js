import API from '../../services/axios';

export const deleteUser = (uid) => {
    return(dispatch,getState,{getfirebase, getFirestore}) => {
        API.post('clients', uid);
    }

}