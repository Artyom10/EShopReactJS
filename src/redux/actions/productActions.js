export const addNewProduct = (newProduct) => {
  return (dispatch, getState, { getFirebase, getFirestore }) => {
    const firestore = getFirestore();
    const storage = getFirebase().storage();

    const infoPhoto = newProduct.photo;
    newProduct.photo = null;

    firestore
      .collection("products")
      .add({
        ...newProduct,
        isBought: false,
      })
      .then((response) => {
        infoPhoto &&
        storage
          .ref(`photos/${response.id}`)
          .put(infoPhoto)
          .on(
            'state_changed',
            snapshot => { },
            error => { console.log(error) },
            () => {
              storage
                .ref('photos')
                .child(response.id)
                .getDownloadURL()
                .then(url => {
                  newProduct.photo = url
                  return firestore
                    .collection('products')
                    .doc(response.id)
                    .update({photo: newProduct.photo})
                })
            }
          )
      })
      .then(() => {
        dispatch({ type: "ADD_PRODUCT", newProduct });
      })
      .catch((err) => {
        dispatch({ type: "ADD_PRODUCT_ERROR", err });
      });
  };
};

export const removeProduct = (targetDeleteProduct) => {
  return (dispatch, getState, { getFirebase, getFirestore }) => {
    console.log(getState());
    const firestore = getFirestore();
    const storage = getFirebase().storage();
    firestore
      .collection("products")
      .doc(targetDeleteProduct)
      .get()
      .then((doc) => {
        firestore
          .collection('products')
          .doc(targetDeleteProduct)
          .delete()
          .then(() => {
            doc.data().photo && 
            storage
              .ref(`photos/${targetDeleteProduct}`)
              .delete()
          })
      })
      .then(() => {
        dispatch({ type: "REMOVE_PRODUCT", targetDeleteProduct });
      })
      .catch((err) => {
        dispatch({ type: "REMOVE_PRODUCT_ERROR", err });
      });

      const buyerId = getState().firebase.auth.uid;
      const firebaseBuyer = getState().firebase;
      let findIndex = null;

      let ratedProducts = firebaseBuyer.profile.valuedProducts;
      ratedProducts.forEach((product,index) => {
        if(product.targetProductRating === targetDeleteProduct){
                findIndex = index;       
        }
      });
      ratedProducts.splice(findIndex,1);

      let findBagsIndex = null;

      let bagsProducts = firebaseBuyer.profile.bags;
      bagsProducts.forEach((product,index) => {
        if(product === targetDeleteProduct){
          findBagsIndex = index
        }
      });
      bagsProducts.splice(findBagsIndex,1);
  
      firestore.collection('users').doc(buyerId).update({
       valuedProducts:   ratedProducts && [...ratedProducts],
       bags: bagsProducts && [...bagsProducts],
      })
      .then(() => {
        dispatch({ type: "REMOVE_PRODUCT" });
      })
      .catch((err) => {
        dispatch({ type: "REMOVE_PRODUCT_ERROR", err });
      });


  };
};

export const editProduct = (targetEditProduct, changes) => {
  debugger;
  return (dispatch, getState, { getFirebase, getFirestore }) => {
    const firestore = getFirestore();
    const storage = getFirebase().storage();
    typeof changes.photo !== 'string' &&  changes.photo
    ?
    storage
      .ref('photos')
      .child(targetEditProduct)
      .put(changes.photo)
      .on(
        'state_changed',
        snapshot => { },
        error => { console.log(error) },
        () => {
          storage
           .ref('photos')
           .child(targetEditProduct)
           .getDownloadURL()
           .then(url => {
             changes.photo = url
             return firestore
               .collection('products')
               .doc(targetEditProduct)
               .update({
                 ...changes
               })
           })
           .then(() => {
            dispatch({ type: "EDIT_PRODUCT", targetEditProduct });
          })
          .catch((err) => {
            dispatch({ type: "EDIT_PRODUCT_ERROR", err });
          });
        }
      )
      :
      firestore
        .collection('products')
        .doc(targetEditProduct)
        .update({
          ...changes
        })
        .then(() => {
          dispatch({ type: "EDIT_PRODUCT", targetEditProduct });
        })
        .catch((err) => {
          dispatch({ type: "EDIT_PRODUCT_ERROR", err });
        });
  };
};

export const buyProduct = (targetProductBuy) => {
  return (dispatch, getState, { getFirebase, getFirestore }) => {
    const firestore = getFirestore();
    const buyerId = getState().firebase.auth.uid;
    const firebaseBuyer = getState().firebase;
    firestore
      .collection("products")
      .doc(targetProductBuy)
      .update({
        isBought: true,
      })
      .then(() => {
        dispatch({ type: "UPDATE_BOUGHT_PRODUCT_STATUS", targetProductBuy });
      })
      .catch((err) => {
        dispatch({ type: "AUPDATE_BOUGHT_PRODUCT_STATUS_ERROR", err });
      });

    firestore
      .collection("users")
      .doc(buyerId)
      .update({
        bags: firebaseBuyer.profile.bags
          ? [...firebaseBuyer.profile.bags, targetProductBuy]
          : [targetProductBuy],
      })
      .then(() => {
        dispatch({ type: "ADD_TO_BOUGHT", targetProductBuy });
      })
      .catch((err) => {
        dispatch({ type: "ADD_TO_BOUGHT_ERROR", err });
      });
  };
};

export const setRating = (targetProductRating, value, certainProduct) => {
  return (dispatch, getState, {getFirebase, getFirestore}) => {
    const firestore = getFirestore();
    const buyerId = getState().firebase.auth.uid;
    const firebaseBuyer = getState().firebase;
  
    if(firebaseBuyer.profile.valuedProducts.some((everyData) => everyData.targetProductRating === targetProductRating)){
    console.log('You have already rated this product, please delete your rate and try again');
    }
    else{
          console.log('This product does not rate yet, we need update product and value')
          firestore.collection('users').doc(buyerId).update({
            valuedProducts: [...firebaseBuyer.profile.valuedProducts, {targetProductRating, value,}],
        })
        .then(() => {
          dispatch({ type: "RATE_PRODUCT", targetProductRating });
        })
        .catch((err) => {
          dispatch({ type: "RATE_PRODUCT_ERROR", err });
        });
    }
  }
}

export const deleteRating = (targetProductDeleteRating) => {
  debugger;
  return (dispatch, getState, {getFirebase, getFirestore}) => {
    const firestore = getFirestore();
    const buyerId = getState().firebase.auth.uid;
    const firebaseBuyer = getState().firebase;
    let findIndex = null;

    let ratedProducts = firebaseBuyer.profile.valuedProducts;
    ratedProducts.forEach((product,index) => {
      if(product.targetProductRating === targetProductDeleteRating){
              findIndex = index;       
      }
    });
    ratedProducts.splice(findIndex,1);

    firestore.collection('users').doc(buyerId).update({
     valuedProducts:   ratedProducts && [...ratedProducts]
    })
    .then(() => {
      dispatch({ type: "DELETE_RATE_PRODUCT", targetProductDeleteRating });
    })
    .catch((err) => {
      dispatch({ type: "DELETE_RATE_PRODUCT_ERROR", err });
    });
  }
}


export const deleteBooked = (productId) => {
  return (dispatch, getState, {getFirebase, getFirestore}) => {
    debugger;
    const firestore = getFirestore();
    const buyerId = getState().firebase.auth.uid;
    const firebaseBuyer = getState().firebase;
    let findIndex = null;

    let bagsProducts = firebaseBuyer.profile.bags;
    bagsProducts.forEach((product,index) => {
      if(product === productId){
              findIndex = index;       
      }
    });
    bagsProducts.splice(findIndex,1);

    firestore.collection('products').doc(productId).update({
      isBought: false,
    })
    .catch((err) => {
      dispatch({type: "DELETE_BAG_PRODUCT_ERROR", err});
    })
    
    firestore.collection('users').doc(buyerId).update({
      bags:   bagsProducts && [...bagsProducts]
     })
     .then(() => {
       dispatch({ type: "DELETE_BAG_PRODUCT" });
     })
     .catch((err) => {
       dispatch({ type: "DELETE_BAG_PRODUCT_ERROR", err });
     });
   }
}