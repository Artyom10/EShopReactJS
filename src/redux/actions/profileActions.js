export const updateProfile = (firstName, secondName) => {
    return (dispatch, getState,{getFirebase, getFirestore}) => {
     const firestore = getFirestore();
     const profile = getState().firebase.auth.uid
     firestore.collection('users').doc(profile).update({
            firstName,
            secondName,
     })
     .then(() => {
         dispatch({type: 'UPDATE_PROFILE'})
     }).catch((err) => {
         dispatch({type: 'UPDATE_PROFILE_ERROR', err})
     })
    }
}

export const removeRequest = (user) => {
    debugger;
    return (dispatch, getState, {getFirebase, getFirestore}) => {
        const firestore = getFirestore();
        const profile = getState().firebase.auth.uid
        firestore.collection('users').doc(profile).update({
            request: true,
        })
        .then(() => {
            dispatch({type: 'UPDATE_PROFILE_REQUEST'})
        }).catch((err) => {
            dispatch({type: 'UPDATE_PROFILE_REQUEST_ERROR', err})
        })
    }
}