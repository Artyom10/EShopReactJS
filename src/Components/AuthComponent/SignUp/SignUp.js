import React, { useState } from "react";
import { Redirect } from "react-router-dom";
import stylesFor from "./SignUp.module.css";
import  signUpSchema  from '../../../validation/signUpSchema';
import { useDispatch, useSelector } from "react-redux";
import { useFormik } from "formik";
import { signUp } from "../../../redux/actions/authActions";

const SignUp = (props) => {
  let [email, setEmail] = useState("");
  let [password, setPassword] = useState("");
  let [firstName, setFirstName] = useState("");
  let [secondName, setSecondName] = useState("");
  let [request, setRequest] = useState(false);
  let [bags, setBags] = useState(null);
  let [valuedProducts, setValuedProducts] = useState([]);

  const handleChange = (e) => {
    if (e.target.id === "email") {
      setEmail(e.target.value);
    } else if (e.target.id === "password") {
      setPassword(e.target.value);
    } else if (e.target.id === "firstName") {
      setFirstName(e.target.value);
    } else if (e.target.id === "secondName") {
      setSecondName(e.target.value);
    }
  };

  const handleSubmit = (e) => {
    e.preventDefault();
    let ourState = {
      email,
      password,
      firstName,
      secondName,
      request,
      bags,
      valuedProducts,
    };
    props.signUp(ourState);
  };

  /*const auth = useSelector((state) => state.firebase.auth);
  const authError = useSelector((state) => state.auth.authError);
  const dispatch = useDispatch();

  const formik = useFormik({
    initialValues: {
      firstName: "",
      secondName: "",
      email: "",
      password: "",

    },
    validationSchema: signUpSchema,
    onSubmit: (values) => {
      dispatch(signUp(values));
    },
  });*/

  const { auth, authError } = props;
  if (auth.uid) return <Redirect to="/" />;
  return (
    <div className="container">
      <div className="container">
        <div className="row  justify-content-center">
          <div className="col-md-5">
            <div className="align-items-center justify-content-center">
              <div
                className={` ${stylesFor.containerSignup} d-flex  align-items-center justify-content-center`}
              >
                <form
                  onSubmit={handleSubmit}
                  className="login-form text-center"
                >
                  <h1 className="mb-5 font-weight-light text-uppercase">
                    Sign Up
                  </h1>
                  <div className="form-group">
                    <input
                      type="text"
                      className={`form-control rounded-pill form-control-lg ${stylesFor.control}`}
                      placeholder="First Name"
                      id="firstName"
                      onChange={handleChange}
                    />
                  </div>
                  <div className="form-group">
                    <input
                      type="text"
                      className={`form-control rounded-pill form-control-lg ${stylesFor.control}`}
                      placeholder="Second Name"
                      id="secondName"
                      onChange={handleChange}
                    />
                  </div>
                  <div className="form-group">
                    <input
                      type="email"
                      className={`form-control rounded-pill form-control-lg ${stylesFor.control}`}
                      placeholder="Email"
                      id="email"
                      onChange={handleChange}
                    />
                  </div>
                  <div className="form-group">
                    <input
                      type="text"
                      className={`form-control rounded-pill form-control-lg ${stylesFor.control}`}
                      placeholder="Password"
                      onChange={handleChange}
                      id="password"
                    />
                  </div>
                  <button
                    type="submit"
                    className="btn mt-4 btn-dark btn-block text-uppercase rounded-pill btn-lg"
                  >
                    Sign Up
                  </button>
                  <p className="mt-3 font-weight-normal">
                    {authError ? <p>{authError}</p> : null}
                  </p>
                </form>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  );
};

export default SignUp;
