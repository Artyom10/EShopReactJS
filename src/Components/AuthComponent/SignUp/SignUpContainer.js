import React from 'react';
import { connect } from 'react-redux';
import {signUp} from '../../../redux/actions/authActions';
import SignUp from './SignUp';

const mapStateToProps = (state) => {
  return {
    authError: state.auth.authError,
    auth: state.firebase.auth,
  }
}

const mapDispatchToProps = (dispatch) => {
  return {
    signUp: (newUser) => dispatch(signUp(newUser))
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(SignUp);
