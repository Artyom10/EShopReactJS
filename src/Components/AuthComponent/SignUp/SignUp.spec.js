import { shallow } from 'enzyme';
import React from 'react';
import {signUp} from '../../../redux/actions/authActions';
import SignUp from './SignUp';

const auth = {
    uid: null,
};

const authRedirect = {
    uid: 111,
};

const props = {
    authError: null,
    signUp,
};

const setUp = (props,authInfo) => shallow(<SignUp {...props} auth={authInfo} />);

describe('LogUp component', () => {
    it('should render LogUp component', () => {
       const component = setUp(props, auth);
       expect(component.debug()).toMatchSnapshot();
    });

    it('should redirect', () => {
        const component = setUp(props, authRedirect);
        expect(component.debug()).toMatchSnapshot();
    });
});