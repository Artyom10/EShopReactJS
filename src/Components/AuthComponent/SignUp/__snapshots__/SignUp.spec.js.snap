// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LogUp component should redirect 1`] = `"<Redirect to=\\"/\\" />"`;

exports[`LogUp component should render LogUp component 1`] = `
"<div className=\\"container\\">
  <div className=\\"container\\">
    <div className=\\"row  justify-content-center\\">
      <div className=\\"col-md-5\\">
        <div className=\\"align-items-center justify-content-center\\">
          <div className=\\" containerSignup d-flex  align-items-center justify-content-center\\">
            <form onSubmit={[Function: handleSubmit]} className=\\"login-form text-center\\">
              <h1 className=\\"mb-5 font-weight-light text-uppercase\\">
                Sign Up
              </h1>
              <div className=\\"form-group\\">
                <input type=\\"text\\" className=\\"form-control rounded-pill form-control-lg control\\" placeholder=\\"First Name\\" id=\\"firstName\\" onChange={[Function: handleChange]} />
              </div>
              <div className=\\"form-group\\">
                <input type=\\"text\\" className=\\"form-control rounded-pill form-control-lg control\\" placeholder=\\"Second Name\\" id=\\"secondName\\" onChange={[Function: handleChange]} />
              </div>
              <div className=\\"form-group\\">
                <input type=\\"email\\" className=\\"form-control rounded-pill form-control-lg control\\" placeholder=\\"Email\\" id=\\"email\\" onChange={[Function: handleChange]} />
              </div>
              <div className=\\"form-group\\">
                <input type=\\"text\\" className=\\"form-control rounded-pill form-control-lg control\\" placeholder=\\"Password\\" onChange={[Function: handleChange]} id=\\"password\\" />
              </div>
              <button type=\\"submit\\" className=\\"btn mt-4 btn-dark btn-block text-uppercase rounded-pill btn-lg\\">
                Sign Up
              </button>
              <p className=\\"mt-3 font-weight-normal\\" />
            </form>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>"
`;
